kani-compiler/src/codegen_cprover_gotoc/overrides/hooks.rs

```diff
@@ -245,6 +245,107 @@ impl<'tcx> GotocHook<'tcx> for Panic {
     }
 }
 
+/// Hook for handling postconditions in function contracts.
+struct Postcondition;
+
+impl<'tcx> GotocHook<'tcx> for Postcondition {
+    fn hook_applies(&self, tcx: TyCtxt<'tcx>, instance: Instance<'tcx>) -> bool {
+        matches_function(tcx, instance, "KaniPostcondition")
+    }
+
+    fn handle(
+        &self,
+        tcx: &mut GotocCtx<'tcx>,
+        _instance: Instance<'tcx>,
+        mut fargs: Vec<Expr>,
+        _assign_to: Place<'tcx>,
+        target: Option<BasicBlock>,
+        span: Option<Span>,
+    ) -> Stmt {
+        assert_eq!(fargs.len(), 1);
+        let cond = fargs.remove(0).cast_to(Type::bool());
```

Comment:
I would suggest you to implement something like:

```rust
struct PostCondition {
  // Should we apply this hook (i.e.: either enforce or check precondition).
  enabled: bool,
}

impl<'tcx> GotocHook<'tcx> for Postcondition {
  fn hook_applies(&self, tcx: TyCtxt<'tcx>, instance: Instance<'tcx>) -> bool {
    self.enabled && matches_function(tcx, instance, "KaniPostcondition")
  }

  fn handle (/*args*/) {
    if tcx.queries.get_enforce_contracts() {
      fargs.add(Expr::string_constant("Post-condition failed");
      Assert::handle(tcx, instance, fargs, assign_to, target, span);
    } else {
      Assume::handle(tcx, instance, fargs, assign_to, target, span);
    }
  }
}
```

Reply:
We want to convert the postcondition as follows - 
1) to an `assert` when a function is being checked against its contract (calling Kani with the `--enforce-contracts` flag)
2) to an `assume` if the function is being replaced with its contract (calling Kani with the `--replace-with-contracts` flag)
3) To an empty statement, otherwise (calling Kani without any flags)

In this snippet, enabled must be set to true if `tcx.queries.get_enforce_contracts() == true` or `tcx.queries.get_replace_with_contracts() == true`. Where can I set this value? Also, how do I ensure that in the last case (that is call Kani normally without any contract checking/replacing flags) the inlined `kani::postcondition` gets converted to an empty line?

Reply:
I've refactored my code to use `Assert.handle(..)` and `Assume.handle(..)` within the same if-else statement.